Added spinner option to pagination widget

orangejenny opened the pull request:
Add the option to show feedback that pagination is "busy" while switching pages. Wanted this to be subtle and not interfere with the list of items being shown. Open to feedback - @dannyroberts @proteusvacuum @dimagi/product might have opinions.

code buddy @millerdev 

![khezi24pbt](https://user-images.githubusercontent.com/1486591/51692693-75ff5280-1fcb-11e9-8004-a926dc34adb0.gif)

orangejenny committed:
Added spinner option to pagination widget

orangejenny committed:
Updated styles for .active links.

These were displaying bootstrap's bright blue when active/focus,
instead of the darker blue HQ uses to show active.

djmore9 commented:
Nice touch.

On Thu, Jan 24, 2019 at 11:31 AM Jenny Schweers <notifications@github.com>
wrote:

> Add the option to show feedback that pagination is "busy" while switching
> pages. Wanted this to be subtle and not interfere with the list of items
> being shown. Open to feedback - @dannyroberts
> <https://github.com/dannyroberts> @proteusvacuum
> <https://github.com/proteusvacuum> @dimagi/product
> <https://github.com/orgs/dimagi/teams/product> might have opinions.
>
> code buddy @millerdev <https://github.com/millerdev>
>
> [image: khezi24pbt]
> <https://user-images.githubusercontent.com/1486591/51692693-75ff5280-1fcb-11e9-8004-a926dc34adb0.gif>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/dimagi/commcare-hq/pull/23090
> Commit Summary
>
>    - Added spinner option to pagination widget
>    - Updated styles for .active links.
>
> File Changes
>
>    - *M* corehq/apps/export/static/export/js/export_list.js
>    <https://github.com/dimagi/commcare-hq/pull/23090/files#diff-0> (15)
>    - *M*
>    corehq/apps/export/templates/export/partials/export_list_controller.html
>    <https://github.com/dimagi/commcare-hq/pull/23090/files#diff-1> (3)
>    - *M* corehq/apps/export/templates/export/partials/loading_exports.html
>    <https://github.com/dimagi/commcare-hq/pull/23090/files#diff-2> (2)
>    - *M* corehq/apps/hqwebapp/static/hqwebapp/js/components/pagination.js
>    <https://github.com/dimagi/commcare-hq/pull/23090/files#diff-3> (5)
>    - *M* corehq/apps/hqwebapp/static/hqwebapp/less/_hq/pagination.less
>    <https://github.com/dimagi/commcare-hq/pull/23090/files#diff-4> (3)
>    - *M* corehq/apps/hqwebapp/templates/hqwebapp/ko_pagination.html
>    <https://github.com/dimagi/commcare-hq/pull/23090/files#diff-5> (5)
>
> Patch Links:
>
>    - https://github.com/dimagi/commcare-hq/pull/23090.patch
>    - https://github.com/dimagi/commcare-hq/pull/23090.diff
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/23090>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdarmReihvhSKdUpVo8aemhIpabLVBwIks5vGd_3gaJpZM4aRPsM>
> .
>


-- 
Devendra More
Director of Product, SaaS
Dimagi, Inc | 585 Massachusetts Ave | Suite 3 | Cambridge, MA 02139
t: 617-649-2214 <(617)%20649-2214> | m: 857-415-9419 <(617)%20649-2214> |
Skype: dmore@dimagi.com
web: http://www.dimagi.com/

dannyroberts reviewed (approved):
Looks good to me!

proteusvacuum commented on corehq/apps/hqwebapp/static/hqwebapp/js/components/pagination.js:
```diff
@@ -9,6 +9,10 @@
  *      totalItems: A knockout observable that returns the total number of items
  *      inlinePageListOnly: Optional. True or false. If true, leave off the "Showing X to Y of Z entries"
  *          text and the associated dropdown.
+ *      showSpinner: Optional. An observable. If provided, the widget will replace the current page number
```
nit: might be nice to call this "showPaginationSpinner" to be explicit

proteusvacuum commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -183,21 +183,23 @@ hqDefine("export/js/export_list", [
         });
         self.showPagination = ko.computed(function () {
-            return !self.isLoading() && !self.hasError() && self.exports().length;
+            return !self.isLoadingPanel() && !self.hasError() && self.exports().length;
```
does "panel" refer to the table / data that is loading?

orangejenny replied:
yeah, panels are these things: https://getbootstrap.com/docs/3.3/components/#panels-heading

proteusvacuum reviewed (approved):
Looks great! Is there a way to prevent the little page box from resizing when the spinner is shown?

millerdev reviewed (approved):
I love subtle and non-obtrusive UI touches like this 🏆

orangejenny committed:
Fixed styles to prevent jumpiness when spinner shows/hides

orangejenny committed:
Added spinner to web users, mobile users, and releases.

Dashboard has different spinner behavior, and data corrections
doesn't use ajax so it doesn't need one.

orangejenny committed:
Merge branch 'master' into jls/pagination-feedback

orangejenny committed:
Fixed lint error

orangejenny commented:
> Is there a way to prevent the little page box from resizing when the spinner is shown?

Is it surprisingly challenging - I futzed with it for a while and didn't find a good solution that works across different pagination sizes (since the whole widget is sized based on font size). Punting on this.

biyeun commented:
🎉 🏆 🍹
